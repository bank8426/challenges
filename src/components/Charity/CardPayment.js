import React,{useState} from 'react'
import { connect } from 'react-redux';
import styled, { keyframes } from 'styled-components';
import * as paymentActions from '../../redux/actions/paymentActions';
import * as messageActions from '../../redux/actions/messageActions';
import { bindActionCreators } from 'redux'
import RadioButton from '../common/RadioButton';
import PropTypes from 'prop-types'

const cardPaymentFadeInKeyframes = keyframes`
  0% {
    opacity: 0;
  }
  100% {
    opacity: 1;
  }
`

export const StyledCardPayment = styled.div`
  position: absolute;
  top: 0px;
  width: 100%;
  height: 100%;
  background-color: rgba(255,255,255,0.95);
  padding: 15% 20px;
  box-sizing: border-box;
  border-radius: 10px;
  opacity: 0;
  -webkit-animation: ${cardPaymentFadeInKeyframes} 0.2s forwards cubic-bezier(0.2, 0.8, 0.2, 1) 0.05s;
  animation: ${cardPaymentFadeInKeyframes} 0.2s forwards cubic-bezier(0.2, 0.8, 0.2, 1) 0.05s;
`;

const StyledPaymentList = styled.div`
  display: grid;
  grid-template-columns: repeat(5,1fr);
  padding: 0 100px;
  @media (max-width: 1280px) {
    padding: 0 50px;
  }
  @media (max-width: 1024px) {
    padding: 0px;
  }
  @media (max-width: 768px) {
    padding: 0 80px;
  }
`;

const StyledHeader = styled.h2`
  text-align: center;
`;

export const StyledCloseButton = styled.span`
  position: absolute;
  top: 10px;
  right: 10px;
  border-radius: 50%;
  cursor: pointer;
  font-size: 20px;
  padding: 1px 8px;
  color: #1b53f2;
  font-weight: bold;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
  -webkit-transition: 0.4s cubic-bezier(0.2, 0.8, 0.2, 1);
  transition: 0.4s cubic-bezier(0.2, 0.8, 0.2, 1);
  &:hover{
    color: white;
    background-color: #1B95F2;
    border-color: #0062cc;
  }
  &:active {
    color: rgba(255, 255, 255, 0.5);
    background-color: #1B95F2;
    border-color: #0062cc;
  }
`;

const StyledPayButton = styled.button`
  display: block;
  margin: 2em auto;
`;
export const CardPayment = ({index,savePayment,addMessage,removeMessageById,currency,onClose}) => {
  const payments = [10, 20, 50, 100, 500];
  const [selectedAmount, setSelectedAmount] = useState(payments[0])
  const handlePay = () => {
    savePayment(index, selectedAmount, currency).catch(error => {
      const id = Date.now();
      addMessage('Something went wrong please try again later.',id ,true);
      setTimeout(function() {
        removeMessageById(id)
      }, 2000);
    })
  }
  return (
    <StyledCardPayment>
      <StyledCloseButton onClick={() => onClose()}>X</StyledCloseButton>
      <StyledHeader>Select the amount to donate ({currency})</StyledHeader>
      <StyledPaymentList>
        {
          payments.map((amount, j) => (
            <RadioButton key={j}
              name={`payment${index}`} 
              handleChange={() => {setSelectedAmount(amount)}} 
              displayMessage={amount} 
              checked={amount===selectedAmount}/>
          ))
        }
      </StyledPaymentList>
      <StyledPayButton className='primary' onClick={() => handlePay()}>Pay</StyledPayButton>
    </StyledCardPayment>
  )
}

CardPayment.defaultProps ={
  index : 0,
  savePayment : () => {},
  addMessage : () => {},
  removeMessageById : () => {},
  currency : '',
  onClose : () => {},
}

CardPayment.propTypes={
  index:PropTypes.number.isRequired,
  savePayment: PropTypes.func.isRequired,
  addMessage: PropTypes.func.isRequired,
  removeMessageById: PropTypes.func.isRequired,
  currency:PropTypes.string.isRequired,
  onClose: PropTypes.func.isRequired,
}

const mapDispatchToProps = (dispatch) => ({ 
  savePayment: bindActionCreators(paymentActions.savePayment, dispatch),
  addMessage: bindActionCreators(messageActions.addMessage, dispatch),
  removeMessageById: bindActionCreators(messageActions.removeMessageById, dispatch),
})

export default connect(null,mapDispatchToProps)(CardPayment)


