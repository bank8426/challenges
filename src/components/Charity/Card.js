import React from 'react'
import styled,{keyframes} from 'styled-components';
import CardPayment from './CardPayment';
import PropTypes from 'prop-types'

const cardFadeInKeyframe = keyframes`
  0% {
    opacity: 0;
    top: 10px;
  }
  100% {
    opacity: 1;
    top: 0px;
  }
`

const StyledCard = styled.div`
  margin: 0 10px 50px;
  background: #FFFFFF;
  box-shadow: 2px 2px 10px rgba(0, 0, 0, 0.25);
  border-radius: 10px 10px 10px 10px;
  position: relative;
  -webkit-animation: ${cardFadeInKeyframe} 3s forwards cubic-bezier(0.2, 0.8, 0.2, 1) ${props => (props.index/8+.02)+'s'};
  animation: ${cardFadeInKeyframe} 3s forwards cubic-bezier(0.2, 0.8, 0.2, 1) ${props => (props.index/8+.02)+'s'};
  opacity: 0;
  top: 10px;
`;

const StyledImg = styled.div`
  width: 650px;
  overflow: hidden;
  background-image: url(${props => props.src});
  height: 300px;
  border-radius: 10px 10px 0px 0px;
  background-repeat: no-repeat;
  background-size: cover;
  background-position: center;

  @media (max-width: 1366px) {
    width: 600px;
    height: 275px;
  }
  @media (max-width: 1280px) {
    width: 480px;
    height: 230px;
  }
  @media (max-width: 1024px) {
    width: 385px;
    height: 195px;
  }
  @media (max-width: 834px) {
    width: 345px;
    height: 175px;
  }
  @media (max-width: 768px) {
    width: 100%;
    height: 300px;
  }
  @media (max-width: 500px) {
    width: 100%;
    height: 250px;
  }
`;

const StyledH3 = styled.h3`
  margin: 1em 0em 1em 1em;
`;

export const StyledButton = styled.button`
  margin: auto 1em auto 0;
  @media (max-width: 834px) and (min-width: 768px) {
    margin: 0.5em 0.5em 0.5em 0;
    }
`;

const StyledNameLabel = styled.div`
  display: grid;
  grid-template-columns: 1fr max-content;
`;

const Card = ({item,selectedCharity,setSelectCharity,index}) => {
  return (
    <StyledCard index={index}>
      <StyledImg src={process.env.IMAGES_PATH + item.image}/>
      <StyledNameLabel>
        <StyledH3>{item.name}</StyledH3>
        <StyledButton className='primary' onClick={() => setSelectCharity(item.id) }>Donate</StyledButton>
      </StyledNameLabel>
      { selectedCharity===item.id ? 
        <CardPayment 
          index={item.id} 
          currency={item.currency} 
          onClose={() => setSelectCharity(0)}/> 
        : null }
    </StyledCard>
  )
}

Card.defaultProps ={
  item : {
    id : 0,
    name : '',
    image : '',
    currency : '',
  },
  selectedCharity : -1,
  setSelectCharity : () => {},
  index : 0,
}

Card.propTypes={
  item: PropTypes.shape({
    id :PropTypes.number.isRequired,
    name:PropTypes.string.isRequired,
    image:PropTypes.string.isRequired,
    currency:PropTypes.string.isRequired,
  }),
  selectedCharity : PropTypes.number.isRequired,
  setSelectCharity : PropTypes.func.isRequired,
  index : PropTypes.number.isRequired,
}


export default Card
