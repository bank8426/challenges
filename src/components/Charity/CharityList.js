import React,{useState} from 'react'
import styled from 'styled-components';
import Card from './Card';
const StyledCharityList = styled.div`
  display: grid;
  grid-template-columns: repeat(2,1fr);
  justify-items: center;
  @media (max-width: 768px) {
    display: block;
  }
`;

const CharityList = ({charities}) => {
  const [selectedCharity,setSelectCharity] = useState(0)
  return (
    <StyledCharityList>
      {
        charities.map((item, i) => 
          <Card key={i} 
            item={item} 
            selectedCharity={selectedCharity} 
            setSelectCharity={setSelectCharity} />)
      }
    </StyledCharityList>
  )
}


export default CharityList
