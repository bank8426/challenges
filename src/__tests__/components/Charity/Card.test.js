import React from 'react'
import Card , {StyledButton} from '../../../components/Charity/Card'
import {StyledCardPayment , StyledCloseButton} from '../../../components/Charity/CardPayment'
import {mount} from 'enzyme'
import 'jest-styled-components'
import configureStore from 'redux-mock-store'
import { Provider } from 'react-redux';

describe('Card with mock Store', () => {
  // create any initial state needed
  const initialState = {}; 
  // here it is possible to pass in any middleware if needed into //configureStore
  const mockStore = configureStore();
  let wrapper;
  let store;
  let mockFn;
  beforeEach(() => {
    mockFn = jest.fn();
    //creates the store with any initial state or middleware needed  
    store = mockStore(initialState)
    wrapper = mount(<Provider store={store}><Card setSelectCharity={mockFn}/></Provider>)
  })

  it('should display Card', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('should handleClick when click Donate button', () => {
    wrapper.find(StyledButton).simulate('click')
    expect(mockFn).toHaveBeenCalledTimes(1)
  })

  it('should display CardPayment when selectedCharity equal id', () => {
    wrapper = mount(<Provider store={store}>
      <Card id={0} selectedCharity={0} setSelectCharity={mockFn}/>
    </Provider>)
    expect(wrapper.find(StyledCardPayment).length).toEqual(1)
  })

  it('should handle onClose when display CardPayment', () => {
    wrapper = mount(<Provider store={store}>
      <Card id={0} selectedCharity={0} setSelectCharity={mockFn}/>
    </Provider>)
    const props = wrapper.find(StyledCloseButton).props()
    const spy = jest.spyOn(props, 'onClick');
    props.onClick()
    expect(spy).toHaveBeenCalledTimes(1)
  })

  it('should have default setSelectCharity', () => {
    wrapper = mount(<Card />)
    const props = wrapper.props()
    const spy = jest.spyOn(props, 'setSelectCharity');
    props.setSelectCharity()
    expect(spy).toHaveBeenCalledTimes(1)
  })
})
