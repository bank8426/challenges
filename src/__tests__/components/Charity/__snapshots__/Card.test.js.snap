// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card with mock Store should display Card 1`] = `ReactWrapper {}`;
